mailing_list/models.py

```diff
@@ -1,33 +1,152 @@
+import logging
+import json
+
 from datetime import datetime
 
 from django.conf import settings
 from django.db import models
+from django.core.cache import cache
+
+from canvas_sdk.methods import enrollments, sections
+from canvas_sdk.utils import get_all_list_data
+from canvas_sdk.exceptions import CanvasAPIError
+
+from icommons_common.canvas_utils import SessionInactivityExpirationRC
+from icommons_common.models import Person
+
+from .listserv_clients.mailgun import MailgunClient as ListservClient
+
+
+SDK_CONTEXT = SessionInactivityExpirationRC(**settings.CANVAS_SDK_SETTINGS)
+
+
+logger = logging.getLogger(__name__)
+
+
+listserv_client = ListservClient()
+
+
+class MailingListManager(models.Manager):
+    def get_or_create_mailing_lists_for_canvas_course_id(self, canvas_course_id, **kwargs):
+        cache_key = settings.CACHE_KEY_LISTS_BY_CANVAS_COURSE_ID % canvas_course_id
+        lists = cache.get(cache_key, {})
+        if not lists:
+            try:
+                canvas_sections = get_all_list_data(SDK_CONTEXT, sections.list_course_sections, canvas_course_id)
+            except CanvasAPIError:
+                logger.exception("Failed to get canvas sections for canvas_course_id %s", canvas_course_id)
+                raise
+
+            mailing_lists_by_section_id = {ml.section_id: ml for ml in MailingList.objects.filter(canvas_course_id=canvas_course_id)}
+
+            defaults = kwargs.get('defaults', {})
+            for s in canvas_sections:
+                section_id = s['id']
+                mailing_list = mailing_lists_by_section_id.get(section_id)
+                if not mailing_list:
+                    create_kwargs = {
+                        'canvas_course_id': canvas_course_id,
+                        'section_id': section_id
+                    }
+                    create_kwargs.update(defaults)
```

Comment:
This is overwriting what's in create_kwargs with what's in defaults, which isn't how I think of defaults working.  If that's the intent, maybe rename defaults to overrides?